RecFCCeeCalorimeter/src/components/CalibrateBenchmarkMethod.cpp

```diff
@@ -0,0 +1,277 @@
+    : GaudiAlgorithm(aName, aSvcLoc), 
+    m_geoSvc("GeoSvc", aName), 
+    m_histSvc("THistSvc", aName),
+    h_parameter(nullptr),
+    m_totalEnergyECal(nullptr),
+    m_totalEnergyHCal(nullptr),
+    m_totalEnergyBoth(nullptr)
+    {
+      declareProperty("ecalBarrelCells", m_ecalBarrelCells, "");
+      declareProperty("hcalBarrelCells", m_hcalBarrelCells, "");
+    }
+
+
+CalibrateBenchmarkMethod::~CalibrateBenchmarkMethod() {}
+
+
+StatusCode CalibrateBenchmarkMethod::initialize() {
+  if (GaudiAlgorithm::initialize().isFailure()) return StatusCode::FAILURE;
+
+  // Check geometry service
+  if (!m_geoSvc) {
+    error() << "Unable to locate Geometry Service! "
+            << "Make sure you have GeoSvc and SimSvc in the right order in the configuration." << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  // Check if readouts exist
+  {
+    bool readoutMissing = false;
+    for (size_t i = 0; i < m_readoutNames.size(); ++i) {
+      auto readouts = m_geoSvc->lcdd()->readouts();
+      if (readouts.find(m_readoutNames.value().at(i)) == readouts.end()) {
+        readoutMissing = true;
+      }
+    }
+    if (readoutMissing) {
+      error() << "Missing readout, exiting!" << endmsg;
+      return StatusCode::FAILURE;
+    }
+  }
+
+  // book histograms for checks
+  m_totalEnergyECal = new TH1F("totalEnergyECal", "Total deposited energy in ECal", 1000, 0, 1.5 * m_energy);
+  if (m_histSvc->regHist("/rec/ecal_total", m_totalEnergyECal).isFailure()) {
+    error() << "Couldn't register histogram" << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  m_totalEnergyHCal = new TH1F("totalEnergyHCal", "Total deposited energy in HCal", 1000, 0, 1.5 * m_energy);
+  if (m_histSvc->regHist("/rec/hcal_total", m_totalEnergyHCal).isFailure()) {
+    error() << "Couldn't register histogram" << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  m_totalEnergyBoth = new TH1F("totalEnergyBoth", "Total deposited energy in ECal+HCal", 1000, 0, 1.5 * m_energy);
+  if (m_histSvc->regHist("/rec/both_total", m_totalEnergyBoth).isFailure()) {
+    error() << "Couldn't register histogram" << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  h_parameter = new TH1F("h_parameter","", 4, 0,4);
+  if (m_histSvc->regHist("/rec/h_parameter", h_parameter).isFailure()) {
+    error() << "Couldn't register histogram" << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  // clear vectors that will be later used for fitting
+  vecEgenerated.clear();
+  vecEinECal_total.clear();
+  vecEinHCal_total.clear();
+  vecEinHCal_firstLayer.clear();
+  vecEinECal_lastLayer.clear();
+
+  return StatusCode::SUCCESS;
+}
+
+
+StatusCode CalibrateBenchmarkMethod::execute() {
+
+  double energyInHCal = 0.;
+  double energyInECal = 0.;
+  double energyInFirstHCalLayer = 0; 
+  double energyInLastECalLayer = 0; 
+  double energyInBoth = 0.;
+
+  std::vector<double> energyInECalLayer(m_numLayersECal, 0.); 
```

Comment:
In this case it is possible to avoid memory allocation inside the event loop (the number of layer is fixed for all events). You could thus declare those vectors outside of the execute loop and just reset entries to zero there to improve performance.

Reply:
fixed now